passes/cmds/partition.cc

```diff
@@ -0,0 +1,501 @@
+/*
+ *  yosys -- Yosys Open SYnthesis Suite
+ *
+ *  Copyright (C) 2020  Alberto Gonzalez <boqwxp@airmail.cc>
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "kernel/yosys.h"
+#include "kernel/cost.h"
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+struct PartitionPass : public Pass {
+	//Constructs a hypergraph from a flat techmapped module, partitions the hypergraph,
+	//recovers the cell partition assignments, creates partition modules with the appropriate
+	//cells, ports, and intra-partition connections, then replaces the original module
+	//with one that instantiates and connects the partition modules.
+	int opt_k, opt_s, opt_imbalance;
+	bool opt_verbose, opt_nocleanup;
+	RTLIL::Module *module, *new_module;
+	RTLIL::Design *design;
+	dict<RTLIL::IdString, int> cell_type_costs;
+	dict<RTLIL::SigBit, int> sigbit_to_edgenum;
+	dict<int, pool<int>> edge_nodes;
+	dict<int, pool<int>> partition_nodes;
+	dict<int, bool> edge_cut;
+	pool<RTLIL::IdString> cut_wires;
+	int node_ctr, edge_ctr, cut_edge_ctr, partitions;
+	std::vector<RTLIL::Module *> partition_modules;
+
+	PartitionPass() : Pass("partition", "split techmapped module into equal-sized parts"), opt_k(-1), opt_s(-1),
+	                  opt_imbalance(5), opt_verbose(false), opt_nocleanup(false), module(nullptr), new_module(nullptr),
+	                  design(nullptr), cell_type_costs(CellCosts::unit_gate_cost()), node_ctr(0), edge_ctr(0),
+	                  cut_edge_ctr(0), partitions(2) { }
+
+	void reset() {
```

Comment:
And this ought to be something like `*this = {};`, or otherwise use the `<Pass>Worker` pattern that's present elsewhere.

Reply:
Oh, good call. `*this = {};` or `*this = PartitionPass();` is much cleaner.

Reply:
Nope, either one of those gives me:
```
ERROR: Assert `first_queued_pass == NULL' failed in kernel/register.cc:135.
```

Reply:
I think that the `reset` function as you wrote it is error-prone in the long run (it requires any updates to class member initializers to be manually reflected in the reset function), and also not very nice conceptually because it effectively makes all of those members global variables, since that's how pass registration works.

I haven't looked into this issue before (I more or less cargo-culted the worker pattern in my passes), but given the above I think it is the reasonable solution here, barring a major redesign of the pass system.

Reply:
Agreed it is a maintenance burden.  Also a good point about initialization costs; I didn't consider that.  Anyway I gave up after 3 min trying to figure out the pass registration system to avoid it, so I'll just add a worker struct.